### PR TITLE
GH-29642: [R] Support for .keep_all = TRUE with distinct()

### PR DESCRIPTION
**Pull request by nealrichardson**
### Rationale for this change

Support a missing feature, just wiring up some stuff from R to Acero, then adding docs and tests.

This is mostly picking up where #13934 started and finishing it out. Thanks @mopcup for the initial lift.

### What changes are included in this PR?

An aggregation binding, some symbol manipulation, and tests. I also cleaned up some dplyr test shims from 2022.

### Are these changes tested?

Yes, though if anyone knows of odd corners in `distinct()` that aren't covered by this, we can add more

### Are there any user-facing changes?

Yes indeed.
* GitHub Issue: #29642

### CODE REVIEW

File: r/tests/testthat/test-dplyr-distinct.R [hunk 1]
```diff
@@ -115,12 +103,57 @@ test_that("across() works in distinct()", {
...
-      collect() %>%
-      arrange(int),
-    tbl
-  )
+  # hash_one prefers to keep non-null values, which is different from .keep_all in dplyr
+  # so we can't compare the result directly
+  expected <- tbl %>%
+    # Drop factor because of #44661:
+    # NotImplemented: Function 'hash_one' has no kernel matching input types
+    #   (dictionary<values=string, indices=int8, ordered=0>, uint8)
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-distinct.R:111]: Is 110-111 the error that someone would get if they tried `distinct(..., .keep_all = TRUE)` with a factor in the table/data.frame? 

We might want to make that a bit nicer / more grokable for folks who might not have the dictionary -> factor knowledge top of mind

File: r/R/dplyr-distinct.R [hunk 2]
```diff
@@ -33,11 +27,28 @@ distinct.arrow_dplyr_query <- function(.data, ..., .keep_all = FALSE) {
     .data <- dplyr::group_by(.data, !!!syms(names(.data)))
   }
 
-  out <- dplyr::summarize(.data, .groups = "drop")
+  if (isTRUE(.keep_all)) {
+    # Note: in regular dplyr, `.keep_all = TRUE` returns the first row's value.
+    # However, Acero's `hash_one` function prefers returning non-null values.
+    # So, you'll get the same shape of data, but the values may differ.
```
Comment (jonkeane) [inline r/R/dplyr-distinct.R:33]: This behavior change is probably either not-impactful, or if folks are relying on it, that is actually a bug in their code. Though it does seem like something we should mention (in docs at least?).

Or maybe with a one-time warning?

**Review by jonkeane — COMMENTED** [review]
Thanks for this! Mostly questions about messaging + conveying some of the nuances

File: r/tests/testthat/test-dplyr-distinct.R (hunk 1 above)
Comment (nealrichardson) [reply r/tests/testthat/test-dplyr-distinct.R:111]: Yeah that's the error message. I'd have to think about how/where best to catch that and translate that to R-speak. As it turns out, dictionary isn't the only unsupported type, it's just the only one we have in this test data frame. I think list types and other non-simple types are also not supported, IIRC from RTFS.

File: r/R/dplyr-distinct.R (hunk 2 above)
Comment (nealrichardson) [reply r/R/dplyr-distinct.R:33]: It is documented on the acero man page, that's the change to arrow-package.R. I'd rather not one-time warning; that's a slippery slope if we were going to be chatty about every subtle difference between how Acero works from dplyr on data.frames.

File: r/tests/testthat/test-dplyr-distinct.R (hunk 1 above)
Comment (nealrichardson) [reply r/tests/testthat/test-dplyr-distinct.R:111]: https://github.com/apache/arrow/issues/44740

**Comment by rkrug** [comment]
Is there any chance to get it merged? I would very much need to use it in my code!

Thanks.

**Review by jonkeane — APPROVED** [review]
My last review should have been an approve with comments rather than just a comment

**Comment by rkrug** [comment]
Thanks. So it will be in the next release?

**Comment by jonkeane** [comment]
It missed the latest release, but once it merges it'll be in the following one. Alternatively, you could use nightly builds starting the day after it merges.

**Pull request merged**